text/0580-destroyables.md

```diff
@@ -0,0 +1,385 @@
+
+The destructor function is passed a single argument, which is the destroyable
+itself. This allows the function to be reused multiple times for many
+destroyables, rather than creating a closure function per destroyable.
+
+```js
+import { registerDestructor } from '@ember/destroyable';
+
+function unregisterResize(instance) {
+  instance.resize.unregister(instance);
+}
+
+class Modal extends Component {
+  @service resize;
+
+  constructor() {
+    this.resize.register(this, this.layout);
+
+    registerDestructor(this, unregisterResize);
+  }
+}
+```
+
+- Registering a destructor on a destroyed object should throw an error.
+- Attempting to register the same destructor multiple times should throw an
+  error.
+
+#### `unregisterDestructor`
+
+Receives a destroyable and a destructor function, and de-associates the
+destructor from the destroyable.
+
+```js
+import { unregisterDestructor } from '@ember/destroyable';
+
+class Modal extends Component {
+  @service modals;
+
+  constructor() {
+    this.modals.add(this);
+
+    this.modalDestructor = registerDestructor(this, () => this.modals.remove(this));
+  }
+
+  @action pinModal() {
+    unregisterDestructor(this, this.modalDestructor);
+  }
+}
+```
+
+- Calling `unregisterDestructor` on a destroyed object should throw an error.
+- Calling `unregisterDestructor` with a destructor that is not associated with
+  the object should throw an error.
+
+#### `destroy`
+
+`destroy` initiates the destruction of a destroyable object. It runs all
+associated destructors, and then destroys all children recursively.
+
+```js
+let obj = {};
+
+registerDestructor(obj, () => console.log('destroyed!'));
+
+destroy(obj); // this will schedule the destructor to be called
+
+// ...some time later, during scheduled destruction
+
+// destroyed!
+```
+
+Destruction via `destroy()` follows these steps:
+
+1. Mark the destroyable such that `isDestroying(destroyable)` returns `true`
+2. Schedule calling the destroyable's destructors
+3. Call `destroy()` on each of the destroyable's associated children
+4. Schedule setting destroyable such that `isDestroyed(destroyable)` returns `true`
+
+This algorithm results in the entire tree of destroyables being first marked as
+destroying, then having all of their destructors called, and finally all being
+marked as `isDestroyed`. There won't be any in between states where some items
+are marked as `isDestroying` while destroying, while others are not.
+
+Calling `destroy` multiple times on the same destroyable is safe. It will not
+throw an error, and will not take any further action.
+
+Calling `destroy` with a destroyable that has no destructors or associated children
+will not throw an error, and will do nothing.
+
+#### `isDestroying`
+
+Receives a destroyable, and returns `true` if the destroyable has begun
+destroying. Otherwise returns false.
+
+```js
+let obj = {};
+
+isDestroying(obj); // false
+destroy(obj);
+isDestroying(obj); // true
+```
```

Comment:
```suggestion
```js
let obj = {};

isDestroying(obj); // false
destroy(obj);
isDestroying(obj); // true

// ...sometime later, after scheduled destruction
isDestroyed(obj); // true
isDestroying(obj); // false
```
```

Will `isDestroying` return `true` or `false` _after_ the destruction has finished (`isDestroyed`)?